libs/auth/src/signed_quick_results_reporting.ts

```diff
@@ -185,57 +214,89 @@ export async function generateSignedQuickResultsReportingUrl(
     signingMachineId,
     precinctSelection,
     pollsState,
+    maxQrCodeLength = MAXIMUM_BYTES_IN_MEDIUM_QR_CODE,
   }: SignedQuickResultsReportingInput,
   configOverride?: SignedQuickResultsReportingConfig
-): Promise<string> {
+): Promise<string[]> {
   const config =
     configOverride ??
     /* istanbul ignore next - @preserve */ constructSignedQuickResultsReportingConfig();
 
   const { ballotHash, election } = electionDefinition;
-  const compressedTally =
-    pollsState === 'polls_closed_final'
-      ? compressAndEncodeTally({
-          election,
-          results,
-          precinctSelection,
-        })
-      : '';
+  let numPartsNeeded = 1; // If we need to paginate this value will be incremented.
   const secondsSince1970 = Math.round(new Date().getTime() / 1000);
 
-  const messagePayload = encodeQuickResultsMessage({
-    ballotHash,
-    signingMachineId,
-    isLiveMode,
-    timestamp: secondsSince1970,
-    compressedTally,
-    pollsState,
-    precinctSelection,
-  });
-  const message = constructPrefixedMessage(QR_MESSAGE_FORMAT, messagePayload);
-  const messageSignature = await signMessage({
-    message: Readable.from(message),
-    signingPrivateKey: config.machinePrivateKey,
-  });
+  while (
+    numPartsNeeded <=
+    // Only paginate if polls are closed and we have results to send.
+    (pollsState === 'polls_closed_final' ? MAX_PARTS_FOR_QR_CODE : 1)
+  ) {
+    const compressedTallies =
+      pollsState === 'polls_closed_final'
+        ? compressAndEncodeTally({
+            election,
+            results,
+            precinctSelection,
+            numParts: numPartsNeeded,
+          })
+        : ['']; // For polls open reports we don't have results to send.
```

Comment:
This has a fair amount of special casing for the `pollsState` both here and in `encodeQuickResultsMessage`. Both functions need to know that a `compressedTally` that is an empty string gets treated specially. While there are presumably tests that enforce this, it indicates to me that this needs to be decomposed so that we can treat the two cases differently using an `if` or `switch` where the containing `while` is in this function. For example, perhaps `encodeQuickResultsMessage` ought to be unaware of exactly what the message is, so instead of taking a `compressedTally` and `pollsState` it simply takes a `message`. Similarly, this whole loop isn't a loop in the `pollsState !== 'polls_closed_final'` state, which is a bit confusing while reading it.